src/parsing.jl

```diff
@@ -238,77 +238,66 @@ end
 # see https://github.com/invenia/LibPQ.jl/issues/33
 _trunc_seconds(str) = replace(str, r"(\.[\d]{3})\d+" => s"\g<1>")
 
-_DEFAULT_TYPE_MAP[:timestamp] = DateTime
-const TIMESTAMP_FORMAT = dateformat"y-m-d HH:MM:SS.s"  # .s is optional here
-function pqparse(::Type{DateTime}, str::AbstractString)
+# Utility function for handling "infinity"  strings for datetime types to reduce duplication
+function _tryparse_datetime_inf(
+    typ::Type{T}, str, f=typ
+)::Union{T, Nothing} where T <: Dates.AbstractDateTime
     if str == "infinity"
         depwarn_timetype_inf()
-        return typemax(DateTime)
+        return f(typemax(DateTime))
     elseif str == "-infinity"
         depwarn_timetype_inf()
-        return typemin(DateTime)
+        return f(typemin(DateTime))
     end
 
-    # Cut off digits after the third after the decimal point,
-    # since DateTime in Julia currently handles only milliseconds, see Issue #33
-    str = replace(str, r"(\.[\d]{3})\d+" => s"\g<1>")
-    return parse(DateTime, str, TIMESTAMP_FORMAT)
+    return nothing
+end
+
+_DEFAULT_TYPE_MAP[:timestamp] = DateTime
+const TIMESTAMP_FORMAT = dateformat"y-m-d HH:MM:SS.s"  # .s is optional here
+function pqparse(::Type{DateTime}, str::AbstractString)
+    parsed = _tryparse_datetime_inf(DateTime, str)
+    isnothing(parsed) || return parsed
+
+    parsed = tryparse(DateTime, str, TIMESTAMP_FORMAT)
+    isnothing(parsed) || return parsed
+
+    return parse(DateTime, _trunc_seconds(str), TIMESTAMP_FORMAT)
 end
 
 # ISO, YMD
 _DEFAULT_TYPE_MAP[:timestamptz] = ZonedDateTime
-const TIMESTAMPTZ_ZDT_FORMATS = (
+const TIMESTAMPTZ_FORMATS = (
     dateformat"y-m-d HH:MM:SSz",
     dateformat"y-m-d HH:MM:SS.sz",
     dateformat"y-m-d HH:MM:SS.ssz",
     dateformat"y-m-d HH:MM:SS.sssz",
 )
-const TIMESTAMPTZ_UTC_FORMATS = (
-    dateformat"y-m-d HH:MM:SS",
-    dateformat"y-m-d HH:MM:SS.s",
-    dateformat"y-m-d HH:MM:SS.ss",
-    dateformat"y-m-d HH:MM:SS.sss",
-)
 
-timestamptz_formats(::Type{ZonedDateTime}) = TIMESTAMPTZ_ZDT_FORMATS
-timestamptz_formats(::Type{UTCDateTime}) = TIMESTAMPTZ_UTC_FORMATS
+function pqparse(::Type{ZonedDateTime}, str::AbstractString)
+    parsed = _tryparse_datetime_inf(ZonedDateTime, str, Base.Fix2(ZonedDateTime, tz"UTC"))
```

Comment:
This adds an additional function abstraction in a couple places so I checked performance between this and master; looks good 👍